### ISSUE TITLE
`hiop+cuda ^cuda@12` failes to build

### ISSUE THREAD

**Issue by cameronrutherford** [LLNL/hiop#670]
https://github.com/spack/spack/issues/40678 - thanks @eugeneswalker for reporting this.

Please refer to spack GitHub issue for full reporting of build failure. From what I can understand, this is the first time I have seen this error.

I speculated that this is code @kswirydo and/or @pelesh know about, but git-blame shows @cnpetra added the lines throwing errors 1+ year ago.

Two questions:
- Anyone have a different hypothesis why this error is happening (apart from `^cuda@12`)?
- Has anyone built with `cuda@12`?

**Comment by pelesh** [comment]
I think @cnpetra is the right person to address this issue. The error comes from CSR matrix class and is related to CUDA-SDK function for CSC->CSR matrix conversion as far as I can see.

**Comment by pelesh** [comment]
> * Anyone have a different hypothesis why this error is happening (apart from `^cuda@12`)?
> * Has anyone built with `cuda@12`?

https://forums.developer.nvidia.com/t/cusparse-csr2csc-agl2-is-missing-at-cuda-12-1/257431

**Comment by cameronrutherford** [comment]
> > * Anyone have a different hypothesis why this error is happening (apart from `^cuda@12`)?
> > * Has anyone built with `cuda@12`?
> 
> https://forums.developer.nvidia.com/t/cusparse-csr2csc-agl2-is-missing-at-cuda-12-1/257431

Thanks @pelesh for finding the docs, and I have edited the issue title to reflect this. Not sure if we can just switch this to the default method without causing other issues - @cnpetra

**Comment by pelesh** [comment]
> > > * Anyone have a different hypothesis why this error is happening (apart from `^cuda@12`)?
> > > * Has anyone built with `cuda@12`?
> > 
> > 
> > https://forums.developer.nvidia.com/t/cusparse-csr2csc-agl2-is-missing-at-cuda-12-1/257431
> 
> Thanks @pelesh for finding the docs, and I have edited the issue title to reflect this. Not sure if we can just switch this to the default method without causing other issues - @cnpetra

As far as I remember one algorithm is fast but inaccurate and the other one is slower but more precise. Changing the algorithm would require a bit of testing to make sure everything still works correctly. This is for @cnpetra to decide.

**Comment by nychiang** [comment]
@cnpetra @cameronrutherford I think this issue has been fixed by PR #671. Please confirm and we can close this issue.

**Comment by cameronrutherford** [comment]
My vote is to close it. I am pretty sure @eugeneswalker was able to spin up successful builds, and is partially documented in [#61 in ExaGO](https://github.com/pnnl/ExaGO/issues/61)

**Comment by eugeneswalker** [comment]
> I think this issue has been fixed by PR #671. Please confirm and we can close this issue.

This fixes it for me, building `hiop@develop ^cuda@12.2.1` works now where previously I hit the error described above:
```
...
==> Installing hiop-develop-yrfkyjxiac5i2vn3qgepcpsvxxwo44oc [20/20]
==> No binary for hiop-develop-yrfkyjxiac5i2vn3qgepcpsvxxwo44oc found: installing from source
==> No patches needed for hiop
==> hiop: Executing phase: 'cmake'
==> hiop: Executing phase: 'build'
==> hiop: Executing phase: 'install'
==> hiop: Successfully installed hiop-develop-yrfkyjxiac5i2vn3qgepcpsvxxwo44oc
  Stage: 2.95s.  Cmake: 15.71s.  Build: 46.65s.  Install: 1.06s.  Post-install: 0.74s.  Total: 1m 7.29s
[+] /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-11.4.0/hiop-develop-yrfkyjxiac5i2vn3qgepcpsvxxwo44oc
```

Thanks!
